crates/sui-types/src/move_package.rs

```diff
@@ -181,15 +182,16 @@ impl MovePackage {
 
     /// Create an upgraded version of the package along with this version's type origin and linkage
     /// tables.
-    pub fn new_upgraded(
-        predecessor: &MovePackage,
+    pub fn new_upgraded<'a, D: IntoIterator<Item = &'a MovePackage>>(
+        &self,
         modules: Vec<CompiledModule>,
         max_move_package_size: u64,
-        dependencies: &[MovePackage],
+        dependencies: D,
     ) -> Result<Self, ExecutionError> {
-        let type_origin_table = Self::build_upgraded_type_origin_table(predecessor, &modules);
-        let mut new_version = SequenceNumber::from(predecessor.version().value());
+        let type_origin_table = Self::build_upgraded_type_origin_table(self, &modules);
+        let mut new_version = SequenceNumber::from(self.version().value());
         new_version.increment();
+        self.check_compatibility(&modules, true, true, false)?;
```

Comment:
Note that we're assuming that friends can only exist within a package so it's fine to break backwards compatibility amongst those. I think this makes sense, but I'm also fine with being a bit more strict right now (full backwards compatibility) and then relaxing it later on. @amnn any opinion here?